### PR TITLE
cortex_m_systick IDLE timer improvement

### PR DESCRIPTION
**Pull request by niedzwiecki-dawid**
A few improvement of SysTick IDLE timer code:
- handle cycle counter overflow
- handle idle timer overflow
- handle very short or 0 sleep time

### CODE REVIEW

**Review by npitre — APPROVED** [review]

**Review by erwango — APPROVED** [review]

**Review by andyross — COMMENTED** [review]
Two nitpicks, not -1's but would be nice to see clarified.

File: drivers/timer/cortex_m_systick.c [hunk 1]
```diff
@@ -402,9 +402,20 @@ void sys_clock_idle_exit(void)
...
 		 * the SysTick missed in idle state.
 		 */
-		measurement_diff_us = idle_timer_us - systick_us;
-		missed_cycles =
-			(sys_clock_hw_cycles_per_sec() * measurement_diff_us) / USEC_PER_SEC;
+		if (idle_timer_us < systick_us) {
+			/* This case is possible, when the time in low power mode is
+			 * very short or 0. SysTick usually has higher measurement
+			 * resolution of than the IDLE timer, thus the measurement of
+			 * passed time since the sys_clock_set_timeout call can be higher.
```
[the remark this answers is not in the document]
Comment (niedzwiecki-dawid) [reply drivers/timer/cortex_m_systick.c:409]: `systick_diff` is calculated from cycle counters `cycle_post_idle` and `cycle_pre_idle`, not SysTick register counter itself. The cycle counters are calculated with the `elapsed` function, which already handles the wrap.

Anyway, the wrap shouldn't happen anyway, because this code is used in case SysTick is stopped in idle. There is a short window when SysTick still works before and after idle, but just before `sys_clock_set_timeout` with idle=true, the `sys_clock_set_timeout` is called with idle=false (with a few more ticks). So the SysTick is reprogrammed with `load` > `min-residency`.

To sum up, I don't see a risk of the wrap and even if it happens, it is handled well.

File: drivers/timer/cortex_m_systick.c [hunk 2]
```diff
@@ -366,17 +366,28 @@ void sys_clock_idle_exit(void)
...
 		idle_timer_diff = idle_timer_post - idle_timer_pre_idle;
 		idle_timer_us = counter_ticks_to_us(idle_timer, idle_timer_diff);
+
+#ifndef CONFIG_CORTEX_M_SYSTICK_64BIT_CYCLE_COUNTER
+		/* Check cycle counter overflow when using uint32_t */
+		if (cycle_pre_idle > cycle_post_idle) {
+			systick_diff = (UINT32_MAX - cycle_pre_idle) + cycle_post_idle + 1;
+		} else {
+			systick_diff = cycle_post_idle - cycle_pre_idle;
+		}
```
[the remark this answers is not in the document]
Comment (niedzwiecki-dawid) [reply drivers/timer/cortex_m_systick.c:395]: OMG, I don't believe how dumb I'm. It is the same code of course. I just copy paste solution from the IDLE timer overflow, but it is a different case. I'll send a new PR.
